config/src/main/java/org/springframework/security/config/annotation/method/configuration/MethodSecurityConfiguration.java

```diff
@@ -0,0 +1,115 @@
+import org.springframework.security.access.method.MethodAuthorizationContext;
+
+/**
+ * Base {@link Configuration} for enabling Spring Security Method Security.
+ *
+ * @author Evgeniy Cheban
+ * @see EnableMethodSecurity
+ */
+@Configuration(proxyBeanMethods = false)
+@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+public class MethodSecurityConfiguration implements ImportAware {
+
+	private final MethodSecurityExpressionHandler defaultMethodSecurityExpressionHandler = new DefaultMethodSecurityExpressionHandler();
+
+	private int advisorOrder;
+
+	@Bean
+	@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+	MethodSecurityAuthorizationManagerAdvisor methodSecurityAdvisor(AuthorizationMethodInterceptor interceptor,
+			AuthorizationManagerBeforeAdvice<MethodInvocation> beforeAdvice,
+			AuthorizationManagerAfterAdvice<MethodInvocation> afterAdvice) {
+		MethodSecurityAuthorizationManagerAdvisor advisor = new MethodSecurityAuthorizationManagerAdvisor(interceptor,
+				beforeAdvice, afterAdvice);
+		advisor.setOrder(this.advisorOrder);
+		return advisor;
+	}
+
+	@Bean
+	@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+	AuthorizationMethodInterceptor authorizationMethodInterceptor(
+			AuthorizationManagerBeforeAdvice<MethodInvocation> beforeAdvice,
+			AuthorizationManagerAfterAdvice<MethodInvocation> afterAdvice) {
+		return new AuthorizationMethodInterceptor(beforeAdvice, afterAdvice);
+	}
+
+	@Bean
+	@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+	AuthorizationManagerBeforeAdvice<MethodInvocation> authorizationManagerBeforeAdvice() {
+		List<AuthorizationManagerBeforeAdvice<MethodAuthorizationContext>> beforeAdvices = new ArrayList<>();
+		beforeAdvices.add(new ExpressionBasedAuthorizationManagerBeforeAdvice(methodSecurityExpressionHandler()));
+		beforeAdvices.add(new SecuredAnnotationAuthorizationManagerBeforeAdvice());
+		return new DelegatingAuthorizationManagerBeforeAdvice(beforeAdvices);
+	}
+
+	@Bean
+	@Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+	AuthorizationManagerAfterAdvice<MethodInvocation> authorizationManagerAfterAdvice() {
+		List<AuthorizationManagerAfterAdvice<MethodAuthorizationContext>> afterAdvices = new ArrayList<>();
+		afterAdvices.add(new ExpressionBasedAuthorizationManagerAfterAdvice(methodSecurityExpressionHandler()));
+		return new DelegatingAuthorizationManagerAfterAdvice(afterAdvices);
+	}
+
+	/**
+	 * Provide a {@link MethodSecurityExpressionHandler} that is registered with the
+	 * {@link ExpressionBasedAuthorizationManagerBeforeAdvice} and the
+	 * {@link ExpressionBasedAuthorizationManagerAfterAdvice}. The default is
+	 * {@link DefaultMethodSecurityExpressionHandler}.
+	 *
+	 * <p>
+	 * Subclasses may override this method to provide a custom
+	 * {@link MethodSecurityExpressionHandler}.
+	 * </p>
+	 * @return the {@link MethodSecurityExpressionHandler} to use
+	 */
+	protected MethodSecurityExpressionHandler methodSecurityExpressionHandler() {
```

Comment:
I know Spring Security did a lot of this in the past, but I'd like to see if this can be changed to be bean-based instead of based on extending a class.

One way that might work would be to do:

```java
@Autowired(required = false)
public void setMethodSecurityExpressionHandler(MethodSecurityExpressionHandler custom) {
    this.methodSecurityExpressionHandler = custom;
}
```

Reply:
If successful, then let's also make the class `final`

Reply:
The `@Configuration` class cannot be final because Spring creates a CGLIB proxy for it. I think we could make it just package private.

Reply:
In fact, setting the `proxyBeanMethods` attribute to `false` allows the `@Configuration` class to be final.

Reply:
Done